src/org/rascalmpl/parser/gtd/SGTDBF.java

```diff
@@ -1423,23 +1408,30 @@ protected AbstractNode parse(AbstractStackNode<P> startNode, URI inputURI, int[]
 					while (!stacksWithNonTerminalsToReduce.isEmpty() || !stacksWithTerminalsToReduce.isEmpty());
 
 					shiftedLevel = true;
+
+					if (!findStacksToReduce()) {
+						if (location == input.length) {
+							EdgesSet<P> startNodeEdgesSet = startNode.getIncomingEdges();
+							int resultStoreId = getResultStoreId(startNode.getId());
+							if (startNodeEdgesSet != null && startNodeEdgesSet.getLastVisitedLevel(resultStoreId) == input.length) {
```

Comment:
The reference `startNode` doesn't seem to change between iterations (the referred object might, though; at least `rootNode` is an alias, and then it's pushed on the stack, so who knows what happens to it...). Anyway, maybe some of the checks here could be done already earlier (outside the loop), if we're confident the content of `startNode` is "stable enough".

Reply:
The `startNode` content is definitely not stable enough to move the checks outside the loop.